components/fxa-client/android/src/main/java/mozilla/appservices/fxaclient/PersistedFirefoxAccount.kt

```diff
@@ -5,54 +5,49 @@
 package mozilla.appservices.fxaclient
 
 import android.util.Log
-import com.sun.jna.Native
-import com.sun.jna.Pointer
-import mozilla.appservices.fxaclient.rust.FxaHandle
-import mozilla.appservices.fxaclient.rust.LibFxAFFI
-import mozilla.appservices.fxaclient.rust.RustError
-import mozilla.appservices.support.native.toNioDirectBuffer
-import java.util.concurrent.atomic.AtomicLong
 import org.json.JSONObject
 
 /**
- * FirefoxAccount represents the authentication state of a client.
+ * PersistedFirefoxAccount represents the authentication state of a client.
+ *
+ * This is a thin wrapper around the `FirefoxAccount` object exposed from Rust.
+ * Its main job is to transparently manage persistence of the account state by
+ * calling the provided callback at appropriate times.
+ *
+ * In future it would be nice to push this logic down into the Rust code,
+ * once UniFFI's support for callback interfaces is a little more battle-tested.
+ *
  */
-class FirefoxAccount(handle: FxaHandle, persistCallback: PersistCallback?) : AutoCloseable {
-    private var handle: AtomicLong = AtomicLong(handle)
+class PersistedFirefoxAccount(inner: FirefoxAccount, persistCallback: PersistCallback?) : AutoCloseable {
+    private var inner: FirefoxAccount = inner
     private var persistCallback: PersistCallback? = persistCallback
 
     /**
-     * Create a FirefoxAccount using the given config.
+     * Create a PersistedFirefoxAccount using the given config.
      *
      * This does not make network requests, and can be used on the main thread.
      */
-    constructor(config: Config, persistCallback: PersistCallback? = null) :
-    this(rustCall { e ->
-        LibFxAFFI.INSTANCE.fxa_new(
-            config.contentUrl,
-            config.clientId,
-            config.redirectUri,
-            config.tokenServerUrlOverride,
-            e
-        )
-    }, persistCallback) {
+    constructor(config: Config, persistCallback: PersistCallback? = null) : this(FirefoxAccount(
+        config.contentUrl,
```

Comment:
This is kind of dumb - we take a `Config` object on the Kotlin side, destructure it into its fields to pass over the FFI, then the Rust side turns it back into its own variant of a `Config` object! That made sense when we had to write the FFI layer by hand, but we should see whether we can nicely expose the Rust `Config` interface to Kotlin and Swift and then just accept a `Config` here in the constructor.